### ISSUE TITLE
memory leak in correct_maxwellian_gyrokinetic

### ISSUE THREAD

**Issue by manauref** [ammarhakim/gkylzero#189]
Currently we are leaking memory in this updater when redirecting pointers for bmag and jacob_tot, specifically we have

```
  up->bmag = mkarr(conf_basis->num_basis, conf_local_ext->volume, use_gpu);
  up->jacob_tot = mkarr(conf_basis->num_basis, conf_local_ext->volume, use_gpu);
  if (use_gpu) {
    gkyl_array_copy(up->bmag, bmag);
    gkyl_array_copy(up->jacob_tot, jacob_tot);
  } else {
    up->bmag = bmag;
    up->jacob_tot = jacob_tot;
  }
```

the use_gpu=false case redirects up->bmag etc after allocating it with mkarr. We should just not allocate it in the use_gpu=false case. Another option is to pass these in the advance method, in which case we'll receive the host or device pointer automatically and no extra allocation will be needed in either case.

**Comment by JunoRavin** [comment]
Technically what's being done here is fine so long as you're careful about what arrays are freed. We do this in a bunch of places in the G0 app like when we create the distribution functions:

```
    s->f = mkarr(app->use_gpu, app->basis.num_basis, s->local_ext.volume);
    s->f1 = mkarr(app->use_gpu, app->basis.num_basis, s->local_ext.volume);
    s->fnew = mkarr(app->use_gpu, app->basis.num_basis, s->local_ext.volume);

    s->f_host = s->f;
    if (app->use_gpu)
      s->f_host = mkarr(false, app->basis.num_basis, s->local_ext.volume);
```

you just have to make sure that you free the host-side array as well.

**Comment by manauref** [comment]
It is not fine.
```bmag``` is an input to the function, so it was allocated elsewhere. Then we allocate another array with ```up->bmag = mkarr```, followed by pointing ```up->bmag``` to ```bmag```. So when you release ```up->bmag``` you are actually releasing ```bmag```. And thus the memory that was allocated wit this ```mkarr``` is leaked.

**Comment by JunoRavin** [comment]
Oh I see your point. The updater is treating bmag the way we treat things like equation objects in hyper_dg. It should only be acquiring a pointer to that array as opposed to now managing the memory of that array.

**Comment by dingyunl** [comment]
> the use_gpu=false case redirects up->bmag etc after allocating it with mkarr. We should just not allocate it in the use_gpu=false case. Another option is to pass these in the advance method, in which case we'll receive the host or device pointer automatically and no extra allocation will be needed in either case.

I see. I moved bmag and jacob_tot from the arguments of the advance method because you suggested that it is better to store them in the updater during the creation of the updater. If this is still the case, maybe it's better to just not allocate them in the use_gpu=false case? 

> I noticed this by reading the warnings about bmag and jacob_tot when compiling on stellar-amd. Compiler warnings should be read and evaluated at least once, do not ignore them unless you know they are harmless.

It's very useful to know this. I will pay attention to the compiler warnings in the future.

**Comment by JunoRavin** [comment]
You could also keep the allocation where it is and change:

    up->bmag = bmag;
    up->jacob_tot = jacob_tot;
    
    to
    
    gkyl_array_set(up->bmag, 1.0, bmag);
    gkyl_array_set(up->jacob_tot, 1.0, jacob_tot);

so that your allocations and frees are independent of whether you are doing a GPU vs. CPU update. You just don't want to set pointers equal to each other; you can always perform the allocation and assign the values of the input to your internal arrays in the updater.

**Comment by dingyunl** [comment]
Got it. Thanks!

**Comment by manauref** [comment]
No, we should not reallocate bmag and jacob_tot inside of this updater. That is wasteful. Only storing a pointer to the quantities allocated externally is sufficient.

**Comment by JunoRavin** [comment]
Sure, but then there shouldn't be an allocation for the GPU version either. It should just be handled like what is done for auxiliary fields for DG Equation objects and you should pass the on_dev pointer to the initialization so that the updater has a pointer to the on_dev version.

**Comment by manauref** [comment]
You can still pass them to the _new method. Doesn’t have to be the advance method.

**Comment by dingyunl** [comment]
> You could also keep the allocation where it is and change:
> 
> ```
> up->bmag = bmag;
> up->jacob_tot = jacob_tot;
> 
> to
> 
> gkyl_array_set(up->bmag, 1.0, bmag);
> gkyl_array_set(up->jacob_tot, 1.0, jacob_tot);
> ```
> 
> so that your allocations and frees are independent of whether you are doing a GPU vs. CPU update. You just don't want to set pointers equal to each other; you can always perform the allocation and assign the values of the input to your internal arrays in the updater.

I have modified my updater according to Jimmy's suggestion. Is it generally true that we should not set pointers equal to each other in our g0 code?

> No, we should not reallocate bmag and jacob_tot inside of this updater. That is wasteful. Only storing a pointer to the quantities allocated externally is sufficient.

I'm afraid I don't entirely understand. What does it mean by "storing a pointer to the quantities allocated externally"? Is this done in the definition of the struct "gkyl_correct_maxwellian_gyrokinetic"? Do pass bmag and Jacob_tot to "_new" or "_advance"?

**Comment by manauref** [comment]
up->bmag = bmag;
up->jacob_tot = jacob_tot;


stores a pointer. Just do this in the _new method. The user will just make
sure to pass the correct (host or device) pointer to _new.

**Comment by JunoRavin** [comment]
It's usually better to pass the array as is and let the updater decide whether it needs the host or device pointer (that way you don't need to have the use_gpu conditional when you call the new method):

  if (use_gpu) {
    up->bmag = bmag->on_dev;
    up->jacob_tot = jacob_tot->on_dev;
  } else {
    up->bmag = bmag;
    up->jacob_tot = jacob_tot;
  }

**Comment by dingyunl** [comment]
> up->bmag = bmag; up->jacob_tot = jacob_tot; stores a pointer. Just do this in the _new method. The user will just make sure to pass the correct (host or device) pointer to _new.
> […](#)

I tried this way before, but it would fail for some cases
`Test test_1x1v_p1...                            [ OK ]
Test test_1x1v_p2...                            [ OK ]
Test test_1x2v_p1...                            Error:
CUDA error: an illegal memory access was encountered

Test test_1x2v_p2...                            Error:
CUDA error: an illegal memory access was encountered

FAILED: 2 of 4 unit tests have failed.`

**Comment by manauref** [comment]
The unit test needs to pass a host gkyl_array or a device gkyl_array for
bmag/jacob_tot to _new depending on use_gpu.

**Comment by JunoRavin** [comment]
This issue is solved in the gk-g0-app branch (https://github.com/ammarhakim/gkylzero/tree/gk-g0-app) now that we utilize the gk geometry object to fetch needed geometric information. Further refactoring of this object will likely be helpful based on the recent Vlasov PR merge that simplifies the correction object (https://github.com/ammarhakim/gkylzero/pull/328).
